FlinkSQL Nebula as Sink

spike-liu opened the pull request:
FlinkSQL: take Nebula as Sink.

Please feel free to review.

BTW, integration test has been tested locally with the docker-compose.yaml as below:

https://github.com/vesoft-inc/nebula-flink-connector/pull/57/files#diff-881de9edd148f6d70ee8c0ebc96cc09268471ba279c5d0b953b0c8abcf9d2e43

![image](https://user-images.githubusercontent.com/9884987/174019990-e3fb5aa2-c23a-4bf0-a4c6-05e678357726.png)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vesoft-inc/nebula-flink-connector?pullRequest=57) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vesoft-inc/nebula-flink-connector?pullRequest=57) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vesoft-inc/nebula-flink-connector?pullRequest=57) it.</sub>

spike-liu commented:
@Nicole00 Thanks for the build approval.

Meanwhile I have fixed the build error caused by check-style and updated the pull request. Please review again.

By the way, this was a mistake and check-style had been enabled in my development environment.

spike-liu commented:
@Nicole00 Would you please spare us some time to review this PR? Our colleagues would like to contribute more codes based on this PR.

QingZ11 commented:
> @Nicole00 We believe you are very busy. Would you please spare us some time to review this PR?
> 
> We are enthusiasm for this project. Our colleagues would like to contribute more codes based on this PR.
> 
> Or is it possible to assign someone else to help us out?


Please wait a moment, I have communicated with Nicole, for she is on a long vacation, so the pr processing is not very timely.

Sorry for our slow reply. And thank you for your patience.

(supported by Google Translate)

Nicole00 commented:
> @QingZ11 it is great to hear from you and thanks for letting us know.

So sorry for the late reply. I'm on maternity leave so not review it in time. We'll review this pr as soon as possible.

Nicole00 commented on connector/src/main/java/org.apache.flink/connector/nebula/table/NebulaDynamicTableFactory.java:
```diff
@@ -80,11 +96,140 @@ public class NebulaDynamicTableFactory implements DynamicTableSourceFactory,
+            throw new IllegalArgumentException(
+                    String.format("The value of '%s' option should not be negative, but is %s.",
+                            EXECUTE_RETRY.key(), config.get(EXECUTE_RETRY)));
+        }
```
maybe we can keep just one parameter for `timeout` and remove the `retry` parameter, because java client 3.0 just exposes `timeout` for `NebulaPoolConfig`.

spike-liu replied:
These are the existing codes. I am fine with review and removed EXECUTE_RETRY related codes.

Nicole00 commented on connector/src/main/java/org.apache.flink/connector/nebula/sink/NebulaVertexTableBatchExecutor.java:
```diff
@@ -0,0 +1,38 @@
+package org.apache.flink.connector.nebula.sink;
```
please append the license

spike-liu replied:
Fixed all the related License codes.

Nicole00 commented on connector/src/main/java/org.apache.flink/connector/nebula/sink/NebulaBatchTableOutputFormat.java:
```diff
@@ -0,0 +1,44 @@
+package org.apache.flink.connector.nebula.sink;
```
useless class?   can we delete it?

Nicole00 commented:
It's a great PR!  thanks again for your work and attention to nebula.

Nicole00 commented:
1. How can we create a graph space using flink sql?
2. And if there's point to define the vid(src id, dst id) data type in the create sql.

Nicole00 commented on connector/src/test/java/org/apache/flink/connector/nebula/sink/AbstractNebulaOutPutFormatITTest.java:
```diff
@@ -0,0 +1,175 @@
+    @Test
+    public void sinkVertexData() throws ExecutionException, InterruptedException {
+        TableEnvironment tableEnvironment =
+                TableEnvironment.create(EnvironmentSettings.inStreamingMode());
```
please make sure the ut works.
We can use docker-compose to start up nebula service and create a space before executing these tests.

spike-liu replied:
Fixed as below:
<img width="559" alt="image" src="https://user-images.githubusercontent.com/9884987/177292634-96635619-562f-48ae-ad23-cac4fc16b6d7.png">

Nicole00 replied:
great tests

spike-liu commented:
> > @QingZ11 it is great to hear from you and thanks for letting us know.
> 
> So sorry for the late reply. I'm on maternity leave so not review it in time. We'll review this pr as soon as possible.

It is our sunny day to hear from you:)

Please forgive us to push you during your leave.

spike-liu commented on connector/src/main/java/org.apache.flink/connector/nebula/table/NebulaDynamicTableFactory.java:
```diff
@@ -114,10 +259,15 @@ public Set<ConfigOption<?>> requiredOptions() {
         set.add(EXECUTE_RETRY);
+        set.add(SRC_INDEX);
+        set.add(DST_INDEX);
+        set.add(RANK_INDEX);
```
Good idea! Done.

spike-liu commented on connector/src/main/java/org.apache.flink/connector/nebula/catalog/factory/NebulaCatalogFactory.java:
```diff
@@ -30,47 +28,29 @@ public class NebulaCatalogFactory implements CatalogFactory {
+        options.add(CONNECT_TIMEOUT);
+        options.add(CONNECT_RETRY);
+        options.add(TIMEOUT);
+        options.add(EXECUTE_RETRY);
```
Fixed.

spike-liu commented on connector/src/test/java/org/apache/flink/connector/nebula/sink/AbstractNebulaOutPutFormatITTest.java:
```diff
@@ -0,0 +1,175 @@
+package org.apache.flink.connector.nebula.sink;
```
Done.

Nicole00 commented:
> > 1. How can we create a graph space using flink sql?
> >    [Spike] Currently Flink SQL only support few statements as below, so no chance for us to create graph space:
> >    https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/dev/table/sql/overview/
> 
> <img alt="image" width="1116" src="https://user-images.githubusercontent.com/9884987/177294011-5461aac4-02c3-466a-82e4-64d7c4d04355.png">
> 
> > 2. And if there's point to define the vid(src id, dst id) data type in the create sql.
> >    [Spike] based on my understanding, srcId and dstId have to be included in insert statements while all the columns included in insert statements have to be included in Create table statement. Hence it seems this is inevitable. Correct me if I am wrong^-^

1. The supported statements in Flink SQL includes CREATE DATABASE, for nebula it's equal to CREATE SPACE.
2. I agree that srcid and dstid should be included in insert statements. but how to define the datatype of vid in CREATE TABLE statement?  like vid STRING or vid BUGINT? In fact, the datatype in CREATE TABLE statement is useless, it's point in CREATE SPACE statement.

Nicole00 commented:
For more clear function and more convenient for review, please submit the source function in another pr? Thanks~

spike-liu commented:
> > > 1. How can we create a graph space using flink sql?
> > >    [Spike] Currently Flink SQL only support few statements as below, so no chance for us to create graph space:
> > >    https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/dev/table/sql/overview/
> > 
> > 
> > <img alt="image" width="1116" src="https://user-images.githubusercontent.com/9884987/177294011-5461aac4-02c3-466a-82e4-64d7c4d04355.png">
> > > 2. And if there's point to define the vid(src id, dst id) data type in the create sql.
> > >    [Spike] based on my understanding, srcId and dstId have to be included in insert statements while all the columns included in insert statements have to be included in Create table statement. Hence it seems this is inevitable. Correct me if I am wrong^-^
> 
> 1. The supported statements in Flink SQL includes CREATE DATABASE, for nebula it's equal to CREATE SPACE.
> 2. I agree that srcid and dstid should be included in insert statements. but how to define the datatype of vid in CREATE TABLE statement?  like vid STRING or vid BUGINT? In fact, the datatype in CREATE TABLE statement is useless, it's point in CREATE SPACE statement.

Yes, you are right. CREATE DATABASE is exactly what you want. However during our user case, production is highly stability sensitive, there is no chance for us to create database during single data stream processing.

spike-liu commented:
> For more clear function and more convenient for review, please submit the source function in another pr? Thanks~

Sorry for the extra complexity. There are some duplicated codes in Source & Sink, so it is hard to separate them.

However I could revert these codes and bring it on in the future if they are still needed.

@Nicole00 It seems there are some duplicated coding works between @liuxiaocs7 and me. How about create a wechat group and let us discuss something there to raise efficiency and conflicts? I have emailed you and included my wechat account.

Nicole00 commented on connector/src/test/java/org/apache/flink/connector/nebula/sink/AbstractNebulaOutPutFormatITTest.java:
```diff
@@ -0,0 +1,270 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
```
please type the date to 2022, the same for other new files.

spike-liu replied:
Fixed.

spike-liu commented:
> > Yes, you are right. CREATE DATABASE is exactly what you want. However in our user case, production is highly stability sensitive, creating database during single data stream processing is forbidden. (not only for nebula, but also for mysql and etc.
> > Maybe we could create an separate issue to track this requirement.
> 
> I totally agree to submit a separate issue and pr for CREATE SPACE. And For flink sql connector, there's no need to forbidden the creating space operation(at most time, the creating space is together with creating tag/edge). Mysql's database is not allowed to be create because the connection to mysql needs database name, but nebula does not.

Issue has been created as below:
https://github.com/vesoft-inc/nebula-flink-connector/issues/62

spike-liu commented:
> Great work！@spike-liu Thanks again for your contribution~

It is my pleasure to work with you:)

liuxiaocs7 commented:
Hello, @spike-liu. I am new to Flink and Nebula, your work inspires me, I have a few questions about the implementation of flink sql as sink.
1. I think now the table is a temporary table, and the table name is `default_catalog`.`default_database`.`table_name`, and in your implementation the table in flink sql create statement is the same as the vertex/edge name in nebula? If that's true,  how to create two table in different nebula graph space with same name?

liuxiaocs7 commented:
Looking forward for your reply, If this is really the problem, I want to try and refine it, because that's the job of OSPP. Thank you.

spike-liu commented:
@liuxiaocs7 replied in https://github.com/vesoft-inc/nebula-flink-connector/issues/58

liuxiaocs7 commented:
Thank you for your reply, I created a comment here but then I noticed that it was a closed pr, I don't know if you could see it in time, so it was also mentioned in #58 . Feel sorry.
